GeosCore/tracer_mod.F90

```diff
@@ -0,0 +1,760 @@
+    TYPE(MetState), INTENT(IN   ) :: State_Met   ! Meteorology state object
+!
+! !INPUT/OUTPUT PARAMETERS:
+!
+    TYPE(ChmState), INTENT(INOUT) :: State_Chm   ! Chemistry state object
+!
+! !OUTPUT PARAMETERS:
+!
+    INTEGER,        INTENT(OUT)   :: RC          ! Failure or success
+!
+! !REMARKS:
+!
+! !REVISION HISTORY:
+!  02 Mar 2023 - M. Sulprizio- Initial version
+!  See https://github.com/geoschem/geos-chem for complete history
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+!
+! !LOCAL VARIABLES:
+!
+    ! Scalars
+    INTEGER                :: I, J,  L
+    INTEGER                :: N, DT
+    REAL(fp)               :: Local_Tally
+    REAL(fp)               :: Total_Area
+    REAL(fp)               :: Total_Spc
+
+    ! SAVEd scalars
+    LOGICAL,  SAVE         :: First = .TRUE.
+
+    ! Strings
+    CHARACTER(LEN=255)     :: ErrMsg,  ThisLoc
+    CHARACTER(LEN=63)      :: OrigUnit
+
+    ! Arrays
+    REAL(fp)               :: Flux(State_Grid%NX,State_Grid%NY)
+    REAL(fp)               :: Mask(State_Grid%NX,State_Grid%NY,State_Grid%NZ)
+
+    ! Pointers to fields in the HEMCO data structure
+    REAL(f4),      POINTER :: O3(:,:,:)   => NULL()
+
+    ! Objects
+    TYPE(Species), POINTER :: SpcInfo
+
+    !========================================================================
+    ! Tracer_Source_Phase begins here!
+    !========================================================================
+
+    ! Initialize
+    RC          = GC_SUCCESS
+    ErrMsg      = ''
+    ThisLoc     = &
+       ' -> at Tracer_Source_Phase (in module GeosCore/tracer_mod.F90)'
+    DT          = GET_TS_CHEM()
+    Local_Tally = 0.0_fp
+    Total_Area  = 0.0_fp
+    Total_Spc   = 0.0_fp
+    Mask        = 1.0_fp
+    Flux        = 0.0_fp
+
+    !=======================================================================
+    ! Convert species units to v/v dry
+    !=======================================================================
+    CALL Convert_Spc_Units( Input_Opt, State_Chm, State_Grid, State_Met, &
+                            'v/v dry', RC, OrigUnit=OrigUnit )
+    IF ( RC /= GC_SUCCESS ) THEN
+       ErrMsg = 'Unit conversion error (kg/kg dry -> v/v dry)'
+       CALL GC_Error( ErrMsg, RC, ThisLoc )
+       RETURN
+    ENDIF
+
+    !========================================================================
+    ! Apply tracer source
+    !========================================================================
+
+    ! Loop over species
+    DO N = 1, State_Chm%nAdvect
+
+       ! Point to the Species Database entry for species N
+       SpcInfo => State_Chm%SpcData(N)%Info
+
+       ! Skip Rn-Pb-Be tracers for now
+       IF ( SpcInfo%Is_RadioNuclide ) CYCLE
+
+       ! Skip if this species does not have a source or if the 
+       ! source is handled by HEMCO
+       IF ( TRIM(SpcInfo%Src_Mode) == 'none'   .or. &
+            TRIM(SpcInfo%Src_Mode) == 'HEMCO' ) CYCLE
```

Comment:
Could we replace these with integer values?  That would spare a string comparison, which is slower.

Reply:
Could be a TODO for later.